src/logger.coffee

```diff
@@ -5,114 +5,237 @@ es = require 'event-stream'
 Lock = require 'rwlock'
 { checkTruthy } = require './lib/validation'
 
-LOG_PUBLISH_INTERVAL = 110
+class NativeLoggerBackend
+	constructor: ->
+		@logPublishInterval = 1000
+		@maxLogLinesPerBatch = 10
+		@maxQueueSize = 100
+		@maxLineLength = 300
+		@publishQueue = []
+		@intervalHandle = null
+		@publishEnabled = false
+		@offlineMode = false
+		@shouldStop = false
+
+	publishLoop: =>
+		startTime = process.hrtime()
+		Promise.try =>
+			return if @offlineMode or !@publishEnabled or @publishQueue.length is 0
+			currentBatch = _.slice(@publishQueue, 0, @maxLogLinesPerBatch)
+			@publishQueue = _.slice(@publishQueue, @maxLogLinesPerBatch)
+			# Silently ignore errors sending logs (just like with pubnub)
+			@apiBinder.logBatch(currentBatch)
+		.catchReturn()
+		.then =>
+			elapsedTime = process.hrtime(startTime)
+			elapsedTimeMs = elapsedTime[0] * 1000 + elapsedTime[1] / 1e6
+			nextDelay = Math.min(@logPublishInterval - elapsedTimeMs, 0)
+			Promise.delay(nextDelay)
+		.then =>
+			if @shouldStop
+				return
+			@publishLoop()
 
-# Pubnub's message size limit is 32KB (unclear on whether it's KB or actually KiB,
-# but we'll be conservative). So we limit a log message to 2 bytes less to account
-# for the [ and ] in the array.
-MAX_LOG_BYTE_SIZE = 30000
-MAX_MESSAGE_INDEX = 9
+	_truncateToMaxLength: (message) =>
+		if message.message.length > @maxLineLength
+			message = _.clone(message)
```

Comment:
I'm not too aware of the context but, do you really need to clone it and preserve the original? Can't you just set the truncated message back into the original object?

Reply:
We can, but I prefer not to - it'd be an unexpected side effect of the function, and this might cause bugs if a caller doesn't expect this in the future